codalab/client/rest_client.py

```diff
@@ -0,0 +1,95 @@
+        self.client_error = client_error
+
+
+class RestClient(object):
+    """
+    Generic class for communicating with a REST API authenticated by OAuth.
+    """
+
+    def __init__(self, base_url):
+        self._base_url = base_url
+
+    def _get_access_token(self):
+        """
+        Should be overriden with a thread-safe method that returns a valid
+        access token.
+        """
+        raise NotImplementedError
+
+    def _make_request(self, method, path, query_params=None, headers=None,
+                      data=None, return_response=False, authorized=True):
+        if headers is None:
+            headers = {}
+
+        if authorized:
+            headers['Authorization'] = 'Bearer ' + self._get_access_token()
+
+        if data is not None and isinstance(data, dict):
+            headers['Content-Type'] = 'application/json'
+            data = json.dumps(data)
+        headers['X-Requested-With'] = 'XMLHttpRequest'
+        if query_params is not None:
+            path = path + '?' + urllib.urlencode(query_params)
+        request = urllib2.Request(self._base_url + path, data=data,
+                                  headers=headers)
+        request.get_method = lambda: method
+        if return_response:
+            return urllib2.urlopen(request)
+        with closing(urllib2.urlopen(request)) as response:
+            if response.headers.get('Content-Type') == 'application/json':
+                response_data = response.read()
+                try:
+                    return json.loads(response_data)
+                except ValueError:
+                    raise RestClientException(
+                        'Invalid JSON: ' + response_data, False)
+
+    def _upload_with_chunked_encoding(self, method, url, query_params, fileobj):
+        # Start the request.
+        parsed_base_url = urlparse.urlparse(self._base_url)
+        if parsed_base_url.scheme == 'http':
+            conn = httplib.HTTPConnection(parsed_base_url.netloc)
+        else:
+            conn = httplib.HTTPSConnection(parsed_base_url.netloc)
+        with closing(conn):
+            conn.putrequest(
+                method, parsed_base_url.path + url + '?' + urllib.urlencode(
+                    query_params))
+
+            # Set headers.
+            conn.putheader('Authorization', 'Bearer ' + self._get_access_token())
+            conn.putheader('Transfer-Encoding', 'chunked')
+            conn.putheader('X-Requested-With', 'XMLHttpRequest')
+            conn.endheaders()
+
+            # Use chunked transfer encoding to send the data through.
+            while True:
+                to_send = fileobj.read(16 * 1024)
+                if not to_send:
+                    break
+                conn.send('%X\r\n%s\r\n' % (len(to_send), to_send))
+            conn.send('0\r\n\r\n')
+
+            # Read the response.
+            response = conn.getresponse()
+            if response.status != 200:
```

Comment:
This will not get properly translated into a UsageError, even though it has an HTTP error code and it should.